worker/uniter/relation/resolver.go

```diff
@@ -72,42 +74,80 @@ func (r *relationsResolver) NextOp(localState resolver.LocalState, remoteState r
+	// Processing them last ensures that upon application removal, the hooks
+	// for other relations can rely on the peer relations still being present.
+	var peerRelations []int
+
+	// Check whether we need to fire a hook for any of the non-peer relations.
+	for relationID, relationSnapshot := range remoteState.Relations {
+		if isPeer, _ := r.stateTracker.IsPeerRelation(relationID); isPeer {
+			peerRelations = append(peerRelations, relationID)
 			continue
 		}
 
-		// If either the unit or the relation are Dying, or the
-		// relation becomes suspended, then the relation should be
-		// broken.
-		var remoteBroken bool
-		if remoteState.Life == life.Dying || relationSnapshot.Life == life.Dying || relationSnapshot.Suspended {
-			relationSnapshot = remotestate.RelationSnapshot{}
-			remoteBroken = true
-			// TODO(axw) if relation is implicit, leave scope & remove.
+		op, err := r.processRelationSnapshot(relationID, relationSnapshot, remoteState, opFactory)
+		if err != nil {
+			if errors.Is(err, resolver.ErrNoOperation) {
+				continue
+			}
+			return nil, errors.Trace(err)
 		}
+		return op, nil
+	}
 
-		// Examine local/remote states and figure out if a hook needs
-		// to be fired for this relation.
-		relState, err := r.stateTracker.State(relationId)
+	// Process the deferred peer relations.
+	for _, relationID := range peerRelations {
+		relationSnapshot := remoteState.Relations[relationID]
+
+		op, err := r.processRelationSnapshot(relationID, relationSnapshot, remoteState, opFactory)
 		if err != nil {
-			//
-			relState = NewState(relationId)
-		}
-		hInfo, err := r.nextHookForRelation(relState, relationSnapshot, remoteBroken)
-		if err == resolver.ErrNoOperation {
-			continue
+			if errors.Is(err, resolver.ErrNoOperation) {
+				continue
+			}
+			return nil, errors.Trace(err)
 		}
-		return opFactory.NewRunHook(hInfo)
+		return op, nil
 	}
 
 	return nil, resolver.ErrNoOperation
 }
 
+func (r *relationsResolver) processRelationSnapshot(
+	relationID int,
+	relationSnapshot remotestate.RelationSnapshot,
+	remoteState remotestate.Snapshot,
+	opFactory operation.Factory,
+) (operation.Operation, error) {
+	if !r.stateTracker.IsKnown(relationID) {
+		r.logger.Infof("unknown relation %d resolving next op", relationID)
+		return nil, resolver.ErrNoOperation
+	} else if isImplicit, _ := r.stateTracker.IsImplicit(relationID); isImplicit {
+		return nil, resolver.ErrNoOperation
+	}
+
+	// If either the unit or the relation are Dying, or the relation
+	// becomes suspended, then the relation should be broken.
+	var remoteBroken bool
+	if remoteState.Life == life.Dying || relationSnapshot.Life == life.Dying || relationSnapshot.Suspended {
+		relationSnapshot = remotestate.RelationSnapshot{}
+		remoteBroken = true
+	}
+
+	// Examine local/remote states and figure out if a hook needs
+	// to be fired for this relation.
+	relState, err := r.stateTracker.State(relationID)
+	if err != nil {
```

Comment:
I understand the only error this returns is a not found, but it would be good to handle this explicitly.